src/Microsoft.Diagnostics.Monitoring.EventPipe/EventPipeStreamProvider.cs

```diff
@@ -28,7 +30,44 @@ public async Task<Stream> ProcessEvents(DiagnosticsClient client, TimeSpan durat
             EventPipeSession session = null;
             try
             {
-                session = await client.StartEventPipeSessionAsync(_sourceConfig.GetProviders(), _sourceConfig.RequestRundown, _sourceConfig.BufferSizeInMB, cancellationToken).ConfigureAwait(false);
+                IEnumerable<EventPipeProvider> providers = _sourceConfig.GetProviders();
+                int bufferSizeInMB = _sourceConfig.BufferSizeInMB;
+                long rundownKeyword = _sourceConfig.RundownKeyword;
+                RetryStrategy retryStrategy = _sourceConfig.RetryStrategy;
+                bool retry = true;
+                while (retry)
+                {
+                    retry = false;
+                    try
+                    {
+                        EventPipeSessionConfiguration config = new(providers, bufferSizeInMB, (rundownKeyword != 0), true, rundownKeyword);
+                        session = await client.StartEventPipeSessionAsync(config, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (UnsupportedCommandException e)
+                    {
+                        if (retryStrategy == RetryStrategy.DropKeywordKeepRundown)
+                        {
+                            Debug.Assert(rundownKeyword != 0);
+                            Debug.Assert(rundownKeyword != EventPipeSession.DefaultRundownKeyword);
```

Comment:
These asserts aren't guaranteed to be true. Its possible we got an UnsupportedCommandException because the runtime is old and doesn't support the CollectTracing2 command.

Reply:
Same as below ... I think the upstream screwed up the retry strategy specification in this case. It shouldn't ask this loop to retry by dropping the keyword if a non-default keyword is not specified at all.